### PR TITLE
[Design] Support Reading from Custom Data Source

### PR DESCRIPTION
**Pull request by terrytangyuan**
Part of https://github.com/sql-machine-learning/elasticdl/issues/1105.

### CODE REVIEW

File: elasticdl/doc/custom_data_source_support_design.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+RecordIO files and reading the index for each file to obtain dictionaries of `{file_path: num_records}`.
+We then use this information to create tasks.
+* ``Worker`` starts a `TaskDataService` that generates a `tf.data.Dataset` instance that contains the actual data.
+The data is fetched in `TaskDataService._gen()` by opening a RecordIO reader to read the data offsets that each task
+we previously created.
+
+In order to support custom data sources in ElasticDL, we propose to make the following changes:
+
+* Rename ``shard_file_name`` in `Task` protobuf definition to `shard_name` so a task is independent with file names and `shard_name` will
+serve as an identifier for the shard.
```
Comment (ywskycn) [inline elasticdl/doc/custom_data_source_support_design.md:36]: So the new `shard` represents a file, or a table, or sth else?

File: elasticdl/doc/custom_data_source_support_design.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    @abstractmethod
+    def read_records(self, task):
+        """This method will be used in `TaskDataService` to read the records based on
+        the information provided for a given task into a Python generator/iterator.
+        """
+        pass
+
+    @abstractmethod
+    def create_task_dispatcher(self):
```
Comment (ywskycn) [inline elasticdl/doc/custom_data_source_support_design.md]: May need to change the func name. At the first glance, I thought users need to handle the creation of training/evaluation tasks. After seeing the following example, it looks like just a dict describing how many records inside each shard. In that case, why not just one AbstractDataReader maps to one dataset, and we don't mix training/evaluation/prediction together?

File: elasticdl/doc/custom_data_source_support_design.md (hunk 1 above)
Comment (terrytangyuan) [reply elasticdl/doc/custom_data_source_support_design.md:36]: It represents a set that contains multiple records, which can be file, table, etc.

File: elasticdl/doc/custom_data_source_support_design.md (hunk 2 above)
Comment (terrytangyuan) [reply elasticdl/doc/custom_data_source_support_design.md]: Good idea! The creation of `TaskDispatcher` object that's returned in this function should not be done by the user. 

Three `AbstractDataReader`s will increase the number of CLI args that we need to add to the client. How about creating three separate methods in this class, e.g. `create_training_shards`, `create_evaluation_shards`, and `create_prediction_shards`?

File: elasticdl/doc/custom_data_source_support_design.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+There are a couple of places in the the current codebase (as of commit [77cc87a](https://github.com/sql-machine-learning/elasticdl/tree/77cc87a90eec54db565849f0ae07d271fd957190))
+that are coupled heavily with RecordIO, for example:
+
+* When we create and dispatch tasks via ``_make_task_dispatcher()``, we rely on scanning directories that contain
+RecordIO files and reading the index for each file to obtain dictionaries of `{file_path: num_records}`.
+We then use this information to create tasks.
+* ``Worker`` starts a `TaskDataService` that generates a `tf.data.Dataset` instance that contains the actual data.
+The data is fetched in `TaskDataService._gen()` by opening a RecordIO reader to read the data offsets that each task
+we previously created.
+
```
Comment (wangkuiyi) [inline elasticdl/doc/custom_data_source_support_design.md:32]: This is a logical gap between these two paragraphs. How do we connect the fact of being tightly-coupled with RecordIO and the requirement of custom data sources?

Is there any problem with being tightly-coupled with RecordIO? Is custom data sources a solution to the problem?

Comment (terrytangyuan) [reply elasticdl/doc/custom_data_source_support_design.md:32]: Thanks for the questions! I've added some brief background and motivation on this to the design doc. Please check it out and continue our discussion here.

File: elasticdl/doc/custom_data_source_support_design.md (hunk 2 above)
Comment (chunyang-wen) [reply elasticdl/doc/custom_data_source_support_design.md]: `create_shards` with a mode defined in `constants.Mode`?

Comment (terrytangyuan) [reply elasticdl/doc/custom_data_source_support_design.md]: @chunyang-wen Good suggestion. Fixed.

**Review by wangkuiyi — APPROVED** [review]

**Pull request merged**